### PR TITLE
rocAL - hardware decoder python support

### PR DESCRIPTION
**Pull request by rrawther**
Added decoder support for pybind and other related changes
Added jupyter notebook for decoder examples

### CODE REVIEW

**Comment by rrawther** [comment]
@kiritigowda : Please ignore the codacy error since it is just warning about the use of assert.
@LakshmiKumar23 : Can you please test this PR

File: apps/image_augmentation/CMakeLists.txt [hunk 1]
```diff
@@ -46,7 +46,7 @@ find_package(OpenCV QUIET)
 find_package(AMDRPP QUIET)
 
 # Application Includes & Libraries
-include_directories(${ROCM_PATH}/${CMAKE_INSTALL_INCLUDEDIR}/mivisionx)
+include_directories(${ROCM_PATH}/${CMAKE_INSTALL_INCLUDEDIR}/mivisionx ${ROCM_PATH}/${CMAKE_INSTALL_INCLUDEDIR}/mivisionx/rocal)
```
Comment (LakshmiKumar23) [inline apps/image_augmentation/CMakeLists.txt:49]: @rrawther Seems like there is a typo. ```${ROCM_PATH}/${CMAKE_INSTALL_INCLUDEDIR}/mivisionx``` is typed twice

File: rocAL/docs/examples/image_processing/decoder.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+from amd.rocal.pipeline import pipeline_def
+from amd.rocal.plugin.pytorch import ROCALClassificationIterator
+import amd.rocal.fn as fn
+import amd.rocal.types as types
+import matplotlib.gridspec as gridspec
+import matplotlib.pyplot as plt
+#%matplotlib inline
+
+seed = 1549361629
+image_dir = "/home/rajy/work/MIVisionX-data/rocal_data/images_jpg/labels_folder"
```
Comment (LakshmiKumar23) [inline rocAL/docs/examples/image_processing/decoder.py:12]: @rrawther Need to remove hardcoded path. Needs to be an arg

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp [hunk 3]
```diff
@@ -527,14 +531,18 @@ rocalJpegCaffe2LMDBRecordSourceSingleShard(
...
 {
     Image* output = nullptr;
     auto context = static_cast<Context*>(p_context);
     try
     {
         bool use_input_dimension = (decode_size_policy == ROCAL_USE_USER_GIVEN_SIZE) || (decode_size_policy == ROCAL_USE_USER_GIVEN_SIZE_RESTRICTED);
         bool decoder_keep_original = (decode_size_policy == ROCAL_USE_USER_GIVEN_SIZE_RESTRICTED) || (decode_size_policy == ROCAL_USE_MAX_SIZE_RESTRICTED);
+        DecoderType decType = DecoderType::TURBO_JPEG; // default
+        if (dec_type == ROCAL_DECODER_OPENCV) decType = DecoderType::OPENCV_DEC;
+        if (dec_type == ROCAL_DECODER_HW_JEPG) decType = DecoderType::HW_JPEG_DEC;
```
Comment (LakshmiKumar23) [inline rocAL/rocAL/source/api/rocal_api_data_loaders.cpp:545]: @rrawther same as above

**Review by LakshmiKumar23 — CHANGES_REQUESTED** [review]

File: apps/image_augmentation/CMakeLists.txt (hunk 1 above)
Comment (rrawther) [reply apps/image_augmentation/CMakeLists.txt:49]: The second one is for rocal include and is needed

File: rocAL/docs/examples/image_processing/decoder.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
+
+from tkinter import W
+from amd.rocal.pipeline import pipeline_def
+from amd.rocal.plugin.pytorch import ROCALClassificationIterator
+import amd.rocal.fn as fn
+import amd.rocal.types as types
+import matplotlib.gridspec as gridspec
+import matplotlib.pyplot as plt
+#%matplotlib inline
```
[the remark this answers is not in the document]
Comment (rrawther) [reply rocAL/docs/examples/image_processing/decoder.py:9]: yes, I think is needed for running on browser. So I have uncommented it

File: rocAL/docs/examples/image_processing/decoder.py (hunk 2 above)
Comment (rrawther) [reply rocAL/docs/examples/image_processing/decoder.py:12]: fixed it

File: rocAL/rocAL/include/api/rocal_api_data_loaders.h [hunk 5]
```diff
@@ -131,17 +133,19 @@ extern "C"  RocalImage  ROCAL_API_CALL rocalSequenceReaderSingleShard(RocalConte
 /// \param decode_size_policy
 /// \param max_width The maximum width of the decoded images, larger or smaller will be resized to closest
 /// \param max_height The maximum height of the decoded images, larger or smaller will be resized to closest
+/// \param rocal_decoder_type Determines the decoder_type, tjpeg or hwdec
 /// \return Reference to the output image
 extern "C"  RocalImage  ROCAL_API_CALL rocalJpegCOCOFileSource(RocalContext context,
                                                         const char* source_path,
-							                            const char* json_path,
+							                                          const char* json_path,
```
[the remark this answers is not in the document]
Comment (rrawther) [reply rocAL/rocAL/include/api/rocal_api_data_loaders.h:140]: fixed it

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp [hunk 6]
```diff
@@ -451,14 +451,18 @@ rocalJpegCaffe2LMDBRecordSource(
...
```
[the remark this answers is not in the document]
Comment (rrawther) [reply rocAL/rocAL/source/api/rocal_api_data_loaders.cpp:465]: fixed it

**Review by LakshmiKumar23 — APPROVED** [review]

File: rocAL/docs/examples/image_processing/decoder.py (hunk 2 above)
Comment (LakshmiKumar23) [reply rocAL/docs/examples/image_processing/decoder.py:11]: @rrawther Seems like this is still there

**Review by LakshmiKumar23 — CHANGES_REQUESTED** [review]
The hardcoded path is still present

**Comment by rrawther** [comment]
@kiritigowda , @LakshmiKumar23 Can this be merged?
The codacy warning can be ignored

**Review by LakshmiKumar23 — APPROVED** [review]
@kiritigowda Can be merged

**Comment by kiritigowda** [comment]
@rrawther & @LakshmiKumar23 fixing a few aspects of CI, will merge to TOT once that is fixed.

**Comment by rrawther** [comment]
@kiritigowda  Can we merge this PR

**Pull request merged**